### PR TITLE
Fix Mutiny schedulers context propagation bug

### LINKED ISSUE #25818: Vert.x Context doesn't get propagated when using Infrastructure.getDefaultWorkerPool()
### Describe the bug

Vert.x Context doesn't get propagated when using `Infrastructure.getDefaultWorkerPool()`

Whenever you use reactive programming and you run Mutiny with `runSubscriptionOn(Infrastructure.getDefaultWorkerPool()` Vert.x context doesn't get propagated, whereas with `Infrastructure.getDefaultExecutor()` does get propagated.

This impacts OpenTelemetry, MDC and any other people/feature that uses Vert.x context for something, since context data is stored inside Vert.x Context and it doesn't get propagated.

### Expected behavior

Using `Infrastructure.getDefaultWorkerPool()` with Mutiny should behave the same as with `Infrastructure.getDefaultExecutor()` and Vert.x Context should be set

### Actual behavior

Using `Infrastructure.getDefaultWorkerPool()` Vert.x Context is null

### How to Reproduce?

_OPENTELEMETRY IS ONLY BEING USED AS AN EXAMPLE AND REPRODUCER, THE ISSUE IS NOT RESTRICTED TO IT_

https://github.com/luneo7/opentelemetry-reproducer

Just run `./mvnw clean test` and you will see that Vert.x Context doesn't get propagated so OTEL span doesn't have the right parent id when using worker pool.

When doing `curl http://localhost:8080/hello-worker-pool` you will see that two distinct spans are created:

Screenshot 2022-05-26 at 13-44-28 Jaeger UI

When doing `curl http://localhost:8080/hello-default-executor` you will see that the spans are created correctly and there are two spans that correlates to each other:

Screenshot 2022-05-26 at 13-46-38 Jaeger UI

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.9.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by jponge**
This is part of a coordinated fix across Quarkus and Mutiny where scheduler wrapping would cause Vert.x context propagation not to be done.

See https://github.com/smallrye/smallrye-mutiny/pull/955

Fixes #25818

### CODE REVIEW

**Review by luneo7 — COMMENTED** [review]
What about the schedule methods? Shouldn't they also propagate Vert.x context?

File: extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java
Comment (jponge) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:31]: I'm trying a variation of your code (we won't add another parameter to `Infrastructure` in Mutiny) but I get an error testing the Mutiny extension:

```
Errors:
[ERROR]   MutinyTest » Runtime java.lang.RuntimeException: io.quarkus.builder.ChainBuildException: No producers for required item class io.quarkus.deployment.builditem.ContextHandlerBuildItem
```

Other than that I have added test cases calling `submit` and `schedule`, context does propagate with your changes to the Vert.x extension.

Comment (jponge) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:31]: Wrapping in an `Optional<...>`

Comment (luneo7) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:31]: Yup, in the mutiny processor the runtimeInit needs to take an `Optional<ContextHandlerBuildItem> contextHandlerBuildItem` and need to update the code when there is no context handler

Comment (luneo7) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:31]: Making it runtimeInit accept optional:

image

Creating the right MutinyScheduler if context handler is not null:
image

Comment (luneo7) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:31]: Pushed it to my branch, if you are basing on something there: https://github.com/quarkusio/quarkus/compare/main...luneo7:mutiny-executors-simple

Comment (jponge) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:31]: I already have something like this locally 👍

Comment (jponge) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:31]: I just let a `null` handler being forwarded, and just call `Runnable::run()` when `null` in `CRSF`

**Comment by jponge** [comment]
a06b438 is an adaptation of @luneo7 code, with added tests.

With these changes, all executor methods (`scheduleXYZ`, `submit`, `execute`) should propagate the Vert.x context, so it indeeds goes beyond the initial fix. Thanks @luneo7 for insisting on these points.

@geoand @cescoffier I'm keeping this PR as a draft right now so we can discuss, I'm not a specialist of the Vert.x extension and there's also a slight change here. I'll eventually squash commits, etc.

File: extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java
Comment (luneo7) [inline extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:32]: Wouldn't `contextHandler.captureContext()` fail if contextHandler is not set (null)? Maybe the switch of contextHandler != null ? new MutinyScheduler(executor) { ... } :  new MutinyScheduler()` and then in the `ContextualRunnableScheduledFuture` you wouldn't need to check if contextHandler is null or not, since it will only be used if it is not null... Maybe that approach is slightly better?

Comment (jponge) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:32]: Ah good catch. I'd rather avoid having 2 classes (plain `MutinyScheduler` vs an anonymous extension. Null checks are cheap, it won't be a performance hog.

**Review by luneo7 — APPROVED** [review]

File: extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/ContextualRunnableScheduledFuture.java
Comment (cescoffier) [inline extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/ContextualRunnableScheduledFuture.java:3]: Avoid star imports.

File: extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java
Comment (cescoffier) [inline extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:5]: Avoid star imports

File: extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/ContextualRunnableScheduledFuture.java
Comment (geoand) [inline extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/ContextualRunnableScheduledFuture.java:3]: Let's avoid start imports please :)

**Comment by geoand** [comment]
I'll leave this one to you and @cescoffier :)

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java
Comment (cescoffier) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:550]: And now the philosophical question: should we distinguish root context and duplicated context? 

I do not believe so, but I think we should consider the question.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java
Comment (jponge) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java:550]: Duplicated contexts do not share the context-local data, so I would say we are running with a _different_ context.

**Comment by gsmet** [comment]
Wondering if we should backport it to 2.10.1.Final as it looks rather annoying? Please add the `triage/backport?` label if I'm not mistaken.

**Comment by jponge** [comment]
(rebased)

**Comment by jponge** [comment]
CI is green, waiting for @cescoffier to do another review

**Comment by jponge** [comment]
Ah... conflicts, great

**Comment by jponge** [comment]
Fixed

**Review by cescoffier — APPROVED** [review]

**Pull request merged**
